src/WebJobs.Script.WebHost/Diagnostics/DiagnosticListenerService.cs

```diff
@@ -0,0 +1,134 @@
+            {
+                if (listener.Name.StartsWith(ScriptConstants.HostDiagnosticSourcePrefix))
+                {
+                    lock (_listenerSubscriptions)
+                    {
+                        IDisposable listenerSubscription = listener.Subscribe(new HostListenerObserver(listener.Name, _logger), IsEventEnabled);
+
+                        if (!_listenerSubscriptions.TryAdd(listener.Name, listenerSubscription))
+                        {
+                            if (_listenerSubscriptions.Remove(listener.Name, out IDisposable existingSubscription))
+                            {
+                                existingSubscription.Dispose();
+                            }
+
+                            _listenerSubscriptions.Add(listener.Name, listenerSubscription);
+                        }
+
+                        _logger.LogInformation("Subscribed to host diagnostic source '{sourceName}'", listener.Name);
+                    }
+                }
+            });
+        }
+
+        private bool IsEventEnabled(string eventName)
+        {
+            return !eventName.StartsWith(ScriptConstants.HostDiagnosticSourceDebugEventNamePrefix) || _debugTraceEnabled;
+        }
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            SubscribeListeners();
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
+
+        public void Dispose()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            _allListenersSubscription?.Dispose();
+            _standbyChangeHandler?.Dispose();
+
+            foreach (var item in _listenerSubscriptions)
+            {
+                item.Value.Dispose();
+            }
+
+            _listenerSubscriptions.Clear();
+            _disposed = true;
+        }
+
+        private class HostListenerObserver : IObserver<KeyValuePair<string, object>>
+        {
+            private readonly string _listenerName;
+            private readonly ILogger _logger;
+
+            public HostListenerObserver(string listenerName, ILogger logger)
+            {
+                _listenerName = listenerName;
+                _logger = logger;
+            }
+
+            public void OnCompleted() { }
+
+            public void OnError(Exception error) { }
+
+            public void OnNext(KeyValuePair<string, object> kvp)
+            {
+                if (kvp.Value is IHostDiagnosticEvent diagnosticEvent)
+                {
+                    diagnosticEvent.LogEvent(_logger);
+                }
+                else
+                {
+                    _logger.LogDebug("Listener '{listener}' emitted event '{eventName}': {payload}", _listenerName, kvp.Key, kvp.Value);
```

Comment:
Yeah, I agree. I actually think it would make sense to just treat this as a source. Thinking about changing `Listener` to `Diagnostic source` here. Thoughts?